compiler/rustc_mir/src/transform/const_goto.rs

```diff
@@ -0,0 +1,129 @@
+//! This pass optimizes the following sequence
+//! ```rust
+//! bb2: {
+//!     _2 = const true;
+//!     goto -> bb3;
+//! }
+//!
+//! bb3: {
+//!     switchInt(_2) -> [false: bb4, otherwise: bb5];
+//! }
+//! ```
+//! into
+//! ```rust
+//! bb2: {
+//!     _2 = const true;
+//!     goto -> bb5;
+//! }
+//! ```
+
+use crate::transform::MirPass;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use rustc_middle::{mir::visit::Visitor, ty::ParamEnv};
+
+use super::simplify::{simplify_cfg, simplify_locals};
+
+pub struct ConstGoto;
+
+impl<'tcx> MirPass<'tcx> for ConstGoto {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        trace!("Running ConstGoto on {:?}", body.source);
+        let param_env = tcx.param_env_reveal_all_normalized(body.source.def_id());
+        let mut opt_finder =
+            ConstGotoOptimizationFinder { tcx, body, optimizations: vec![], param_env };
+        opt_finder.visit_body(body);
+        let should_simplify = !opt_finder.optimizations.is_empty();
+        for opt in opt_finder.optimizations {
+            let terminator = body.basic_blocks_mut()[opt.bb_with_goto].terminator_mut();
+            let new_goto = TerminatorKind::Goto { target: opt.target_to_use_in_goto };
+            debug!("SUCCESS: replacing `{:?}` with `{:?}`", terminator.kind, new_goto);
+            terminator.kind = new_goto;
+        }
+
+        // if we applied optimizations, we potentially have some cfg to cleanup to
+        // make it easier for further passes
+        if should_simplify {
+            simplify_cfg(body);
+            simplify_locals(body, tcx);
+        }
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for ConstGotoOptimizationFinder<'a, 'tcx> {
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        let mut bailer = || {
+            match terminator.kind {
+                TerminatorKind::Goto { target } => {
+                    // We only apply this optimization if the last statement is a const assignment
+                    let last_statement =
+                        self.body.basic_blocks()[location.block].statements.last()?;
+
+                    match &last_statement.kind {
+                        StatementKind::Assign(box (place, Rvalue::Use(op))) => {
+                            let _const = op.constant()?;
+                            // We found a constant being assigned to `place`.
+                            // Now check that the target of this Goto switches on this place.
+                            let target_bb = &self.body.basic_blocks()[target];
+                            if !target_bb.statements.is_empty() {
```

Comment:
We could also check whether this just contains `StorageDead` statements and copy them over to the current block. I think this can happen when the switch that calls the current block uses a local in the switch statement and that local isn't used anymore after the switch statement. Or is that `StorageDead` already guaranteed to be in the current block?

Reply:
I don't know if `StorageDead` is guaranteed to be in a specific block. It sounds tricky to assume this. The current behaviour is a bit conservative, but also much easier to validate that it is correct. The past few MIR opt unsound bugs have been primarily related to situations where `StorageDead` were not being handled correct. I therefore suggest that I add a FIXME  saying that the current behaviour is conservative. What do you think?

Reply:
Sounds good, you're right, we should be more careful here

Reply:
Fixme added